plutus-core/plutus-core/test/Evaluation/Spec.hs

```diff
@@ -1,11 +1,247 @@
+                  _                                      -> error "genArg: encountered non-Constant term"
+          pure
+            ( Constant () (someValueOf (DefaultUniApply DefaultUniProtoList uniElem) valElems),
+              show argStrings
+            )
+        _ -> error "genArg: encountered non-Constant term"
+  -- Descend upon `Opaque`
+  | Just [_, SomeTypeRep tr'] <- matchTyCon @Opaque tr = genArg tr'
+  -- Descend upon `SomeConstant`
+  | Just [_, SomeTypeRep tr'] <- matchTyCon @SomeConstant tr = genArg tr'
+  -- In the current implementation, all type variables are instantiated
+  -- to `Integer` (TODO: change this).
+  | Just _ <- matchTyCon' "PlutusCore.Builtin.Polymorphism" "TyVarRep" tr =
+      pure . mkArg =<< genInteger
+  | otherwise =
+      error $
+        "genArg: I don't know how to generate builtin arguments of this type: " <> show tr
+
+mkArg :: (Contains DefaultUni a, Show a) => a -> (Term, String)
+mkArg a = (Constant () (someValue a), show a)
+
+-- | If the given `TypeRep`'s `TyCon` is @con@, return its type arguments.
+matchTyCon :: forall con a. (Typeable con) => TypeRep a -> Maybe [SomeTypeRep]
+matchTyCon tr = if con == con' then Just args else Nothing
+  where
+    (con, args) = splitTyConApp (SomeTypeRep tr)
+    con' = typeRepTyCon (typeRep @con)
+
+-- | If the given `TypeRep`'s `TyCon` matches the given module and name, return its type arguments.
+matchTyCon' :: forall a. String -> String -> TypeRep a -> Maybe [SomeTypeRep]
+matchTyCon' modu name tr = if modu == modu' && name == name' then Just args else Nothing
+  where
+    (con, args) = splitTyConApp (SomeTypeRep tr)
+    modu' = tyConModule con
+    name' = tyConName con
+
+type family Head a where
+  Head (x ': xs) = x
+
+----------------------------------------------------------
+-- Generators
+
+genInteger :: Gen Integer
+genInteger = fromIntegral @Int64 <$> Gen.enumBounded
+
+genByteString :: Gen BS.ByteString
+genByteString = Gen.utf8 (Range.linear 0 100) Gen.enumBounded
+
+genText :: Gen Text
+genText = Gen.text (Range.linear 0 100) Gen.enumBounded
+
+genData :: Int -> Gen Data
+genData depth =
+  Gen.choice $
+    [genI, genB]
+      <> [ genRec | depth > 1, genRec <-
+                                 [ genList (depth - 1),
+                                   genMap (depth - 1),
+                                   genConstr (depth - 1)
+                                 ]
+         ]
+
+genI :: Gen Data
+genI = I <$> genInteger
+
+genB :: Gen Data
+genB = B <$> genByteString
+
+genList :: Int -> Gen Data
+genList depth = List <$> Gen.list (Range.linear 0 5) (genData (depth - 1))
+
+genMap :: Int -> Gen Data
+genMap depth =
+  Map
+    <$> Gen.list
+      (Range.linear 0 5)
+      ((,) <$> genData (depth - 1) <*> genData (depth - 1))
+
+genConstr :: Int -> Gen Data
+genConstr depth =
+  Constr <$> genInteger
+    <*> Gen.list
+      (Range.linear 0 5)
+      (genData (depth - 1))
```

Comment:
This is good for the purpose of this PR, but here's a general question: @michaelpj we probably want to have some universal generator for `Data` providing various interesting values etc?

Reply:
I'm pretty sure there is one already. Not sure where it is off the top of my head, but a quick grep should turn it up.

Reply:
There's a `genBoundedData` but it returns `QuickCheck.Gen` rather than `Hedgehog.Gen`, and it takes two additional parameters `imem` and `bmem` which I'd rather not having to provide.

Reply:
There's an [issue](https://input-output.atlassian.net/browse/SCP-3653) to write a better generator.  There is one in the cost modelling code [here](https://github.com/input-output-hk/plutus/blob/b30235933273f6024dc3fbb18543f3b9a02b8126/plutus-core/cost-model/budgeting-bench/Generators.hs#L132) but it's not ideal, and it's not totally clear to me how to write a better one: we need a generator that samples the space of Data objects "uniformly", but I'm not even sure what that means in this context.  See also the comments in #4480, where I had some difficulty coming up with decent samples.

Reply:
Argh, the hedgehog/quickcheck issue again :/